### ISSUE TITLE
[ENH]: Accept positional arguments in FuncFormatter

### ISSUE THREAD

**Issue by stevenlis** [matplotlib/matplotlib#26792]
### Problem

I'm struggling to format an x tick.  I am attempting to format the time axis in subplots so that the xtick label appears only after the start of the data. I tried a custom formatting function, but it appears that the `date_min` cannot be passed for unknown reasons.

```python
import pandas as pd
import numpy as np
from datetime import date

import matplotlib.pyplot as plt
from matplotlib.dates import MonthLocator, num2date
from matplotlib.ticker import FuncFormatter

xs = [pd.date_range(f'{y}-07-01', '2021-12-31', freq='M')
      for y in range(2016, 2019)]
ys = [np.random.rand(len(x)) for x in xs]

fig, axs = plt.subplots(3, 1, figsize=(10, 8))
for ax, x, y in zip(axs, xs, ys):
    ax.plot(x, y)

    # Custom formatting function
    date_min = x[0].date()
    def custom_date_formatter(x, pos):
        dt = num2date(x)
        if dt.date() < date_min:
            return None
        elif dt.month == 1:
            return dt.strftime('%Y')
        else:
            return dt.strftime('%b').upper()

    ax.xaxis.set_major_locator(MonthLocator((1, 4, 7, 10)))
    ax.xaxis.set_major_formatter(FuncFormatter(custom_date_formatter))
    ax.tick_params(axis='x', labelsize=8)

    ax.set_xlim(date(2016, 7, 1))
```
I ended up with the same tick label pattern across the axis in the following plot.
![HpdKd](https://github.com/matplotlib/matplotlib/assets/33796896/a960750b-bb0b-4ab1-9009-00334cba2d3e)

What I would expect the xticklabel to be
```text
JUL - OCT - 2017 - APR - JUL - OCT - 2018 .....
..... JUL - OCT - 2018 - APR - JUL - OCT - 2019 .....
............ JUL - OCT - 2019 - APR - JUL - OCT - 2020 .....
```
![d5c9d6e2-aac4-4aa6-9bb6-e124a4ae9272](https://github.com/matplotlib/matplotlib/assets/33796896/226aa540-4452-4915-be18-3a03d272b672)

 

### Proposed solution

It would be great if we could pass positional arguments. You can define a function like
```python
def custom_date_formatter(x, pos, arg_1, arg_2):
```
Then pass it in FuncFormatter like
```python
FuncFormatter(custom_date_formatter, args=(arg_1_value, arg_2_value,))
```
This is very similar to pandas' [apply](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.apply.html).

Currently, I don't know if there is a solution for the issue I'm facing.

**Comment by ksunden** [comment]
You should be able to do what you want with [functools.partial](https://docs.python.org/3/library/functools.html#functools.partial) (though note that order of positional args is I think reversed from what you are suggesting, so you may need to do the additional args as keyword args)

This is a stdlib, more generic solution, that if it works I would think is preferable to rolling our own version of it.

**Comment by stevenlis** [comment]
@ksunden Thank you. Tbh, I'm not familiar with `functools.partial` at all. How should I edit my function? I tried the following but it doesn't seem to work:
```python
    # Custom formatting function
    def custom_date_formatter(x, pos, date_min):
        dt = num2date(x)
        if dt.date() < date_min:
            return None
        elif dt.month == 1:
            return dt.strftime('%Y')
        else:
            return dt.strftime('%b').upper()

    ax.xaxis.set_major_locator(MonthLocator((1, 4, 7, 10)))
    ax.xaxis.set_major_formatter(
        FuncFormatter(
            partial(custom_date_formatter, date_min=x[0].date())
        )
    )
    ax.tick_params(axis='x', labelsize=8)
```

**Comment by jklymak** [comment]
You have a typo `custom_date_formatter(x, pos, date_min=None):`

**Comment by tacaswell** [comment]
As for why this is happening, see the answer to https://stackoverflow.com/questions/13355233/python-lambda-closure-scoping

TL;DR due to late binding in the closure all of the functions see the last value of `date_min`

```python
import pandas as pd
import numpy as np
from datetime import date

import matplotlib.pyplot as plt
from matplotlib.dates import MonthLocator, num2date
from matplotlib.ticker import FuncFormatter

xs = [pd.date_range(f'{y}-07-01', '2021-12-31', freq='M')
      for y in range(2016, 2019)]
ys = [np.random.rand(len(x)) for x in xs]

fig, axs = plt.subplots(3, 1, figsize=(10, 8))
for ax, x, y in zip(axs, xs, ys):
    ax.plot(x, y)

    # Custom formatting function
    date_min = x[0].date()
    # SEE HERE!!
    def custom_date_formatter(x, pos, date_min=date_min):  # <- note default value
        dt = num2date(x)
        if dt.date() < date_min:
            return None
        elif dt.month == 1:
            return dt.strftime('%Y')
        else:
            return dt.strftime('%b').upper()

    ax.xaxis.set_major_locator(MonthLocator((1, 4, 7, 10)))
    ax.xaxis.set_major_formatter(FuncFormatter(custom_date_formatter))
    ax.tick_params(axis='x', labelsize=8)

    ax.set_xlim(date(2016, 7, 1))
plt.show()

```

also works because the default value on evaluated at function definition time and hence each function will see its own version.

------

I am going to close this because there are more general ways to get the desired functionality using language features or stdlib tools.

**Comment by stevenlis** [comment]
@tacaswell Thank you. It makes sense, but I wonder if it would be a good idea to add an example in the documentation. Perhaps a more generic and simple example could be provided to demonstrate how to pass a default value to `FuncFormatter`.
